### PR TITLE
App http channel middlewares

### PR DESCRIPTION
**Pull request by mcandeia**
# Description

This PR adds support to app http channel middlewares.

Currently, our http middlewares don't execute in a service-to-service invocation context, and some of our middlewares are intented to be executed regardless of used dapr protocol.

App Http Channel middlewares solve this by always executing in a context of the app channel protocol, meaning they are also applied when making service-to-service invocations.

The middleware component itself still the same as used for dapr HTTP API, it could be reused with the caveat of having a mapping from fasthttp.Request to fasthttp.RequestCtx. although they are very similar they don't provide same set of features.

> in draft until having an e2e test using service-to-service invocation + middleware execution

## Issue reference

Please reference the issue this PR will close: #4475 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#2778
* [x] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#2778
* [x] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#2778

### CODE REVIEW

**Review by artursouza — APPROVED** [review]

**Comment by artursouza** [comment]
@mcandeia Please, resolve conflicts.

File: pkg/channel/http/http_channel.go [hunk 1]
```diff
@@ -235,24 +240,29 @@ func (h *Channel) invokeMethodV1(ctx context.Context, req *invokev1.InvokeMethod
 	diag.DefaultHTTPMonitoring.ClientRequestStarted(ctx, verb, req.Message().Method, int64(len(req.Message().Data.GetValue())))
 	startRequest := time.Now()
 
-	// Send request to user application
-	resp := fasthttp.AcquireResponse()
```
Comment (yaron2) [inline pkg/channel/http/http_channel.go:239]: Why is this removed?

Comment (mcandeia) [reply pkg/channel/http/http_channel.go:239]: Sadly, this was a side-effect of converting from fasthttp.Request to fasthttp.RequestCtx,

One solution was simply getting one fasthttp.Response from the pool and set on the fasthttp.RequestCtx, but it's not possible because responses doesn't support copying, and fasthttp.RequestCtx requires a value rather than pointer,

I can use ctx.Response.CopyTo(&resp) but as per [documentation](https://pkg.go.dev/github.com/valyala/fasthttp#Response.CopyTo) it doesn't copy the body stream

image

Also, if the idea is to provide response recycling and less GC pressure I think that copying won't work anyways,

Any suggestion on this @yaron2 ?

Comment (mcandeia) [reply pkg/channel/http/http_channel.go:239]: I'm completely aligned if you think this is a blocker for this PR. Especially if there have been previous experiences that resulted in more memory usage and increased latency (related to GC pauses)

Comment (yaron2) [reply pkg/channel/http/http_channel.go:239]: Currently I'd put this as a blocker but it doesn't mean it'll remain that way unless fixed. We need to be able to measure the performance hit here. An immediate solution would be to revert to the current (and more performant) request construction if no middleware is present.

Comment (mcandeia) [reply pkg/channel/http/http_channel.go:239]: Great idea @yaron2. Applied your suggestions

> We need to be able to measure the performance hit here

Yes, I'll probably ask `@tanvigour` for help to look at the graphs we already have that the memory consumption/p99 latency increases

**Comment by mcandeia** [comment]
@yaron2 @artursouza 

I’m afraid I don’t have bandwidth to get this ready until 1.9.

Do you guys agree on adding a feature flag for this - and launch as an alpha feature, and let the performance impact measurement as a prerequisite to become stable?

From the user perspective there’s no breaking change involved for future releases

**Comment by artursouza** [comment]
> @yaron2 @artursouza
> 
> I’m afraid I don’t have bandwidth to get this ready until 1.9.
> 
> Do you guys agree on adding a feature flag for this - and launch as an alpha feature, and let the performance impact measurement as a prerequisite to become stable?
> 
> From the user perspective there’s no breaking change involved for future releases

This feature should go in 1.9 - putting behind a feature flag makes sense. @yaron2 would this unblock this to be in 1.9?

**Comment by yaron2** [comment]
> > @yaron2 @artursouza
> > I’m afraid I don’t have bandwidth to get this ready until 1.9.
> > Do you guys agree on adding a feature flag for this - and launch as an alpha feature, and let the performance impact measurement as a prerequisite to become stable?
> > From the user perspective there’s no breaking change involved for future releases
> 
> This feature should go in 1.9 - putting behind a feature flag makes sense. @yaron2 would this unblock this to be in 1.9?

If the feature flag guarantees no performance hit then I am ok with this for 1.9

**Comment by mcandeia** [comment]
Perfect, I think it’s already covered by the changed code, if no middleware are specified the actual behavior is kept.

I’ll add a feature for extra caution, so both should be satisfied, pipeline specified and flag enabled, thanks

**Comment by mcandeia** [comment]
Done @artursouza

**Comment by yaron2** [comment]
> > @yaron2 @artursouza
> > I’m afraid I don’t have bandwidth to get this ready until 1.9.
> > Do you guys agree on adding a feature flag for this - and launch as an alpha feature, and let the performance impact measurement as a prerequisite to become stable?
> > From the user perspective there’s no breaking change involved for future releases
> 
> This feature should go in 1.9 - putting behind a feature flag makes sense. @yaron2 would this unblock this to be in 1.9?

Ok, to be clear: what is missing in your eyes from making this ready?

Also, if we retain the pooled responses if no middleware is specified, then I don't think we need a feature flag. But that depends on the answer to my first question :)

**Comment by mcandeia** [comment]
From my perspective this feature is done - when considering it for alpha/preview release,

What's missing: Measure the performance hit when using it - if no middleware is specified nothing changes from the current behavior as @yaron2 suggested

> Also, if we retain the pooled responses if no middleware is specified, then I don't think we need a feature flag. But that depends on the answer to my first question :)

Well, if so, I can revert my previous commit.

**Review by artursouza — APPROVED** [review]

**Comment by ItalyPaleAle** [comment]
I just discovered this PR while working on a merge conflict for #5170. In there, fasthttp is being removed from the HTTP channel (and it will have to be removed from Dapr entirely in #4979), so we may need to figure out what to do with this if we want to merge streaming in 1.10.

**Comment by msfussell** [comment]
Why do we add features like this without any explanation to the "why" of this, even if this may seem important?

**Comment by yaron2** [comment]
> @artursouza @mcandeia @yaron2 Why do we add features like this without any explanation to the "why" of this, even if this may seem important? And then there is no communication on this until a doc issue arrives. Looks like this is not just a surprise to me either.

The rational and why are all in the referenced issue: https://github.com/dapr/dapr/issues/4475
